buchschloss/gui2/actions.py

```diff
@@ -1,303 +1,436 @@
+        if label is not None:
+            label['wraplength'] = WRAPLENGTH
+        return label
+
+
+class SearchForm(BaseForm):
+    """Add search options (and/or) + exact matching and adapt widgets"""
+    # PyCharm seems not to inherit the hint...
+    all_widgets: 'dict[str, dict[Any, Optional[tuple]]]' = {
+        'search_mode': {FormTag.SEARCH: (
+            RadioChoices, [(c, utils.get_name(c)) for c in ('and', 'or')], {})},
+        'exact_match': {FormTag.SEARCH: Checkbox},
+    }
 
-
-# noinspection PyDefaultArgument
-def generic_formbased_action(form_type, form_cls, callback,
-                             form_options={}, fill_data=None,
-                             post_init=lambda f: None,
-                             do_reset=True):
-    # TODO: make form_type an enum
-    """perform a generic action
-        Arguments:
-            form_type: 'new', 'edit' or 'search': used for the form group
-            form_cls: the form class (subclass of misc.tkstuff.forms.Form)
-            callback: the function to call on form submit with form data as keyword arguments
-            form_options: optional dict of additional options for the form
-            fill_data: optional callable taking the input of the first form field
-                and returning a dict of data to fill into the form
-            post_init: called after creation and placement of the form
-                with the form instance as argument
-            do_reset: boolean indicating whether to call app.reset after form submission"""
-    form_options_ = {
-        k: {'groups': v} for k, v in {
-            'new': [forms.ElementGroup.NEW],
-            'edit': [forms.ElementGroup.EDIT],
-            'search': [forms.ElementGroup.SEARCH],
-            None: [],
-        }.items()
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        for ws in cls.all_widgets.values():
+            if ws[None] is not None:
+                w, *a, kw = ws[None]
+                if issubclass(w, (Checkbox, OptionsFromSearch)):
+                    kw = {**kw, 'allow_none': True}
+                    ws.setdefault(FormTag.SEARCH, (w, *a, kw))
+
+    def get_data(self):
+        """ignore empty data"""
+        if self.tag is FormTag.SEARCH:
+            return {k: v
+                    for k, v in super().get_data().items()
+                    if v or isinstance(v, bool)}
+        else:
+            return super().get_data()
+
+    def validate(self):
+        """ignore errors from empty widgets"""
+        errors = super().validate()
+        # NOTE: the password entry will raise ValueError if the passwords don't
+        # match, but it shouldn't be used in searches anyway.
+        # All other widgets shouldn't raise exceptions in .get()
+        if self.tag is FormTag.SEARCH:
+            data = self.get_data()
+            for k in errors.keys() - data.keys():
+                del errors[k]
+        return errors
```

Comment:
This does kind of conflict with "return None on error" (at least in Entry+DropdownChoices, possibly elsewhere)